controllers/storagecluster/cephfilesystem.go

```diff
@@ -19,59 +20,73 @@ type ocsCephFilesystems struct{}
 
 // newCephFilesystemInstances returns the cephFilesystem instances that should be created
 // on first run.
-func (r *StorageClusterReconciler) newCephFilesystemInstances(initData *ocsv1.StorageCluster) ([]*cephv1.CephFilesystem, error) {
+func (r *StorageClusterReconciler) newCephFilesystemInstances(initStorageCluster *ocsv1.StorageCluster) ([]*cephv1.CephFilesystem, error) {
 	ret := &cephv1.CephFilesystem{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      generateNameForCephFilesystem(initData),
-			Namespace: initData.Namespace,
+			Name:      generateNameForCephFilesystem(initStorageCluster),
+			Namespace: initStorageCluster.Namespace,
 		},
 		Spec: cephv1.FilesystemSpec{
 			MetadataPool: cephv1.PoolSpec{
-				Replicated:    generateCephReplicatedSpec(initData, "metadata"),
-				FailureDomain: initData.Status.FailureDomain,
+				Replicated:    generateCephReplicatedSpec(initStorageCluster, "metadata"),
+				FailureDomain: initStorageCluster.Status.FailureDomain,
 			},
 			MetadataServer: cephv1.MetadataServerSpec{
 				ActiveCount:   1,
 				ActiveStandby: true,
-				Placement:     getPlacement(initData, "mds"),
-				Resources:     defaults.GetDaemonResources("mds", initData.Spec.Resources),
+				Placement:     getPlacement(initStorageCluster, "mds"),
+				Resources:     defaults.GetDaemonResources("mds", initStorageCluster.Spec.Resources),
 				// set PriorityClassName for the MDS pods
 				PriorityClassName: openshiftUserCritical,
 			},
 		},
 	}
 
-	if initData.Spec.StorageProfiles == nil {
-		// standalone deployment will not have storageProfile, we need to
-		// define default dataPool, if storageProfile is set this will be
-		// overridden.
-		ret.Spec.DataPools = []cephv1.NamedPoolSpec{
-			{
-				PoolSpec: cephv1.PoolSpec{
-					DeviceClass:   generateDeviceClass(initData),
-					Replicated:    generateCephReplicatedSpec(initData, "data"),
-					FailureDomain: initData.Status.FailureDomain,
+	// Load all StorageProfile objects in the StorageCluster's namespace
+	storageProfiles := &ocsv1.StorageProfileList{}
+	spErr := r.Client.List(r.ctx, storageProfiles, &client.ListOptions{Namespace: initStorageCluster.GetNamespace()})
+	if spErr != nil {
+		r.Log.Error(spErr, fmt.Sprint(ocsv1.StorageProfileListError))
+		return nil, fmt.Errorf("%s", ocsv1.StorageProfileListError)
+	}
+
+	if storageProfiles == nil {
```

Comment:
This needs to be removed because we need the provider mode logic to run even if we don't find any storage profiles. Mainly as error handling for storage profile reconciliation (including not finding any) should go inside the if that capture the provider mode